external/storm-hdfs/src/main/java/org/apache/storm/hdfs/spout/HdfsSpout.java

```diff
@@ -0,0 +1,702 @@
+      classType.getConstructor(FileSystem.class, Path.class, Map.class);
+      return;
+    } catch (ClassNotFoundException e) {
+      LOG.error(readerType + " not found in classpath.", e);
+      throw new IllegalArgumentException(readerType + " not found in classpath.", e);
+    } catch (NoSuchMethodException e) {
+      LOG.error(readerType + " is missing the expected constructor for Readers.", e);
+      throw new IllegalArgumentException(readerType + " is missing the expected constuctor for Readers.");
+    }
+  }
+
+  @Override
+  public void ack(Object msgId) {
+    LOG.trace("Ack received for msg {} on spout {}", msgId, spoutId);
+    if(!ackEnabled) {
+      return;
+    }
+    MessageId id = (MessageId) msgId;
+    inflight.remove(id);
+    ++acksSinceLastCommit;
+    tracker.recordAckedOffset(id.offset);
+    commitProgress(tracker.getCommitPosition());
+    if(fileReadCompletely && inflight.isEmpty()) {
+      markFileAsDone(reader.getFilePath());
+      reader = null;
+    }
+    super.ack(msgId);
+  }
+
+  private boolean canCommitNow() {
+
+    if( commitFrequencyCount>0 &&  acksSinceLastCommit >= commitFrequencyCount ) {
+      return true;
+    }
+    return commitTimeElapsed.get();
+  }
+
+  @Override
+  public void fail(Object msgId) {
+    LOG.trace("Fail received for msg id {} on spout {}", msgId, spoutId);
+    super.fail(msgId);
+    if(ackEnabled) {
+      HdfsUtils.Pair<MessageId, List<Object>> item = HdfsUtils.Pair.of(msgId, inflight.remove(msgId));
+      retryList.add(item);
+    }
+  }
+
+  private FileReader pickNextFile()  {
+    try {
+      // 1) If there are any abandoned files, pick oldest one
+      lock = getOldestExpiredLock();
+      if (lock != null) {
+        LOG.debug("Spout {} now took over ownership of abandoned FileLock {}", spoutId, lock.getLockFile());
+        Path file = getFileForLockFile(lock.getLockFile(), sourceDirPath);
+        String resumeFromOffset = lock.getLastLogEntry().fileOffset;
+        LOG.info("Resuming processing of abandoned file : {}", file);
+        return createFileReader(file, resumeFromOffset);
+      }
+
+      // 2) If no abandoned files, then pick oldest file in sourceDirPath, lock it and rename it
+      Collection<Path> listing = HdfsUtils.listFilesByModificationTime(hdfs, sourceDirPath, 0);
+
+      for (Path file : listing) {
+        if( file.getName().endsWith(inprogress_suffix) ) {
+          continue;
+        }
+        if( file.getName().endsWith(ignoreSuffix) ) {
+          continue;
+        }
+        lock = FileLock.tryLock(hdfs, file, lockDirPath, spoutId);
+        if( lock==null ) {
+          LOG.debug("Unable to get FileLock for {}, so skipping it.", file);
+          continue; // could not lock, so try another file.
+        }
+        LOG.info("Processing : {} ", file);
+        Path newFile = renameSelectedFile(file);
+        return createFileReader(newFile);
+      }
+
+      return null;
+    } catch (IOException e) {
```

Comment:
Shouldnt the filelock get released in case of IOException and any other Exception?